lib/linked_list.rb

```diff
@@ -19,99 +19,228 @@ def initialize
+        marker = marker.next
+      end
+      return min
     end
 
 
     # method that returns the length of the singly linked list
     def length
-      raise NotImplementedError
+      length = 0
+      current = @head    
+      until current.nil?
+        length += 1
+        current = current.next
+      end
+      return length
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
     def get_at_index(index)
-      raise NotImplementedError
+      return nil if index > self.length || index < 0
+      curr = @head
+
+      index.times do
+        curr = curr.next
+      end
+
+      return curr.data
     end
 
     # method to print all the values in the linked list
     def visit
-      raise NotImplementedError
+      return nil if @head.nil?
+      curr = @head
+
+      until curr.nil?
+        p curr.data
+        curr = curr.next
+      end
     end
 
     # method to delete the first node found with specified value
     def delete(value)
-      raise NotImplementedError
+      curr = @head
+      previous = nil
+
+      return nil if curr.nil?
+
+      until curr.data == value
+        if curr.nil?
+          return nil
+        else
+          previous = curr
+          curr = curr.next
+        end
+      end
+
+      if previous == nil
+        @head = curr.next
+      else
+        previous.next = curr.next
+      end
     end
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
     def reverse
-      raise NotImplementedError
+      return if @head.nil? || @head.next.nil?
+
+      curr = @head
+      prev = nil
+
+      until curr.nil?
+        temp = curr.next
+        curr.next = prev
+        prev = curr
+        curr = temp
+      end
     end
 
 
     ## Advanced Exercises
     # returns the value at the middle element in the singly linked list
     def find_middle_value
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      slow = @head
+      fast = @head.next
+
+      until fast.nil?
+        slow = slow.next
+        fast = fast.next.next
+      end
+
+      return slow.data
     end
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
     def find_nth_from_end(n)
-      raise NotImplementedError
+      return nil if @head.nil? || self.length < n
+      
+      slow = @head
+      fast = @head
+
+      (n - 1).times do
+        fast = fast.next
+      end 
+
+      until fast.next.nil?
+        slow = slow.next
+        fast = fast.next
+      end
+
+      return slow.data
     end
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
     def has_cycle

```

Comment:
This doesn't quite work as the cycle doesn't have to connect the head to the tail, it could be a circle starting toward the end.  Like if the last node points to the 2nd to last node.